File: 05_OO/polymorphism/polimorphism.py
class Message:
    def __init__(self, formatter):
        self.formatter = formatter
    
    def send(self):
        return self.formatter.format()

class HelloWorldFormatter:
    def format(self):
        return "Hello, World!"

class GoodMorningFormatter:
    def format(self):
        return "Good Morning!"

# HelloWorldFormatterを使ってメッセージを送る
message1 = Message(HelloWorldFormatter())
print(message1.send())  # "Hello, World!"

# GoodMorningFormatterを使ってメッセージを送る
message2 = Message(GoodMorningFormatter())
print(message2.send())  # "Good Morning!"

File: 05_OO/polymorphism/polimorphism.cpp
#include <iostream>
#include <string>

// interface
class Formatter{
public:
    virtual std::string format() = 0;
};

class Message{
private:
    Formatter* formatter;

public:
    Message(Formatter* fmt) : formatter(fmt) {}

    void send(){
        std::cout << formatter->format() << std::endl;
    }
};

class HellWorldFormatter : public Formatter{
public:
    std::string format() override{
        return "Hell World";
    }
};

class GoodMorningFormatter : public Formatter{
public:
    std::string format() override{
        return "Good Morning";
    }
};

int main(){
    Message message1(new HellWorldFormatter());
    message1.send();

    Message message2(new GoodMorningFormatter());
    message2.send();

    return 0;
}